xds/internal/balancer/gracefulswitch/gracefulswitch_test.go

```diff
@@ -0,0 +1,1354 @@
+}
+
+func (mb1 *mockBalancer1) Close() {
+	mb1.closeCh.Send(struct{}{})
+}
+
+// waitForClientConnUpdate verifies if the mockBalancer1 receives the
+// provided ClientConnState within a reasonable amount of time.
+func (mb1 *mockBalancer1) waitForClientConnUpdate(ctx context.Context, wantCCS balancer.ClientConnState) error {
+	ccs, err := mb1.ccsCh.Receive(ctx)
+	if err != nil {
+		return err
+	}
+	gotCCS := ccs.(balancer.ClientConnState)
+	if diff := cmp.Diff(gotCCS, wantCCS, cmpopts.IgnoreFields(resolver.State{}, "Attributes")); diff != "" {
+		return fmt.Errorf("received unexpected ClientConnState, diff (-got +want): %v", diff)
+	}
+	return nil
+}
+
+// waitForSubConnUpdate verifies if the mockBalancer1 receives the provided
+// SubConn update before the context expires.
+func (mb1 *mockBalancer1) waitForSubConnUpdate(ctx context.Context, wantSCS subConnWithState) error {
+	scs, err := mb1.scStateCh.Receive(ctx)
+	if err != nil {
+		return err
+	}
+	gotSCS := scs.(subConnWithState)
+	if !cmp.Equal(gotSCS, wantSCS, cmp.AllowUnexported(subConnWithState{}, testutils.TestSubConn{})) {
+		return fmt.Errorf("received SubConnState: %+v, want %+v", gotSCS, wantSCS)
+	}
+	return nil
+}
+
+// waitForResolverError verifies if the mockBalancer1 receives the provided
+// resolver error before the context expires.
+func (mb1 *mockBalancer1) waitForResolverError(ctx context.Context, wantErr error) error {
+	gotErr, err := mb1.resolverErrCh.Receive(ctx)
+	if err != nil {
+		return err
+	}
+	if gotErr != wantErr {
+		return fmt.Errorf("received resolver error: %v, want %v", gotErr, wantErr)
+	}
+	return nil
+}
+
+// waitForClose verifies that the mockBalancer1 is closed before the context
+// expires.
+func (mb1 *mockBalancer1) waitForClose(ctx context.Context) error {
+	if _, err := mb1.closeCh.Receive(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (mb1 *mockBalancer1) updateState(state balancer.State) {
+	mb1.cc.UpdateState(state)
+}
+
+func (mb1 *mockBalancer1) newSubConn(addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
+	return mb1.cc.NewSubConn(addrs, opts)
+}
+
+func (mb1 *mockBalancer1) updateAddresses(sc balancer.SubConn, addrs []resolver.Address) {
+	mb1.cc.UpdateAddresses(sc, addrs)
+}
+
+func (mb1 *mockBalancer1) removeSubConn(sc balancer.SubConn) {
+	mb1.cc.RemoveSubConn(sc)
+}
+
+type bb2 struct{}
+
+func (bb2) Build(cc balancer.ClientConn, opts balancer.BuildOptions) balancer.Balancer {
+	return &mockBalancer2{
+		ccsCh:         testutils.NewChannel(),
+		scStateCh:     testutils.NewChannel(),
+		resolverErrCh: testutils.NewChannel(),
+		closeCh:       testutils.NewChannel(),
+		cc:            cc,
+	}
+}
+
+func (bb2) Name() string {
+	return balancerName2
+}
+
+// mockBalancer is a fake balancer used to verify different actions from
+// the gracefulswitch. It contains a bunch of channels to signal different events
+// to the test.
+type mockBalancer2 struct {
```

Comment:
`type mockBalancer2 mockBalancer1` maybe?
If not, then `type mockBalancer2 struct { mockBalancer1 }`

Reply:
Went with Menghan's suggestion of two builders which return same balancer type.